rust/worker/src/server.rs

```diff
@@ -137,454 +135,230 @@ impl WorkerServer {
         let system = self.clone_system()?;
-        let result = orchestrator.run(system).await.map_err(|e| {
-            tracing::error!("Error running orchestrator: {}", e);
-            Status::new(
-                e.code().into(),
-                format!("Error running orchestrator: {}", e),
-            )
-        })?;
-
-        let mut output = Vec::with_capacity(result.records.len());
-        for record in result.records {
-            let metadata = if request.include_metadata {
-                let mut meta = record.metadata.unwrap_or_default();
-
-                // The transport layer assumes the document exists in the metadata
-                // with the special key "chroma:document"
-                if let Some(doc) = record.document {
-                    meta.insert("chroma:document".to_string(), MetadataValue::Str(doc));
-                }
-                Some(chroma_proto::UpdateMetadata::from(meta))
-            } else {
-                None
-            };
+        let knn_inner = knn.into_inner();
+        let scan = knn_inner
+            .scan
+            .ok_or(Status::invalid_argument("Invalid Scan Operator"))?;
+        let (fetch_log_operator, fetch_segment_operator) = self.decompose_proto_scan(scan)?;
+        let filter = knn_inner
+            .filter
+            .ok_or(Status::invalid_argument("Invalid Filter Operator"))?;
+        let knn = knn_inner
+            .knn
+            .ok_or(Status::invalid_argument("Invalid Scan Operator"))?;
+
+        let sieve = KnnFilterOrchestrator::new(
+            self.blockfile_provider.clone(),
+            dispatcher.clone(),
+            self.hnsw_index_provider.clone(),
+            // TODO: Make this configurable
+            1000,
+            fetch_log_operator,
+            fetch_segment_operator,
+            filter.try_into()?,
+        );
 
-            output.push(chroma_proto::MetadataEmbeddingRecord {
-                id: record.id,
-                metadata,
-            });
-        }
+        let matching_records = match sieve.run(system.clone()).await {
+            Ok(output) => output,
+            Err(KnnError::EmptyCollection) => {
+                return Ok(Response::new(to_proto_knn_batch_result(
+                    once(Default::default())
+                        .cycle()
+                        .take(knn.embeddings.len())
+                        .collect(),
+                )?));
+            }
+            Err(e) => {
+                return Err(Status::new(e.code().into(), e.to_string()));
+            }
+        };
+
+        let projection = knn_inner
+            .projection
+            .ok_or(Status::invalid_argument("Invalid Projection Operator"))?;
+        let knn_projection = KnnProjectionOperator::try_from(projection)?;
 
-        // This is an implementation stub
-        let response = chroma_proto::QueryMetadataResponse { records: output };
-        Ok(Response::new(response))
+        match try_join_all(
```

Comment:
maybe use buffers_unordered (think about ordering) since `try_join_all` internally switches between large and small modes and this could obfuscate

Reply:
replaced `try_join_all` with `buffered` which preserve future order